repart: Allow combining CopyBlocks= and CopyFiles=

DaanDeMeyer opened the pull request:
Let's allow the combination of these two options. When used, repart
will first try to apply the CopyBlocks= behavior. If that's not possible,
it falls back to the CopyFiles= behavior.

This is a first step in being able to also use the partition definition
files shipped in the image to build the image in mkosi instead of having
a separate set of repart definition files to build the image.

bluca reviewed (changes requested):
Needs documentation updates? Also can you add some test coverage?

DaanDeMeyer commented:
Updated documentation, added tests and also added a new --oem option to only include OEM partitions which are partitions marked with `Oem=yes` or `FactoryReset=no` if `Oem=` is not configured.

DaanDeMeyer commented:
Alternative names for --oem could be --factory or --vendor. Not sure which one we prefer.

bluca reviewed (approved):
I think I prefer 'vendor' a bit more than 'oem', as the former can also mean distro vendor while the latter is usually used to refer to hardware vendors only, but it's not too important

yuwata commented on src/partition/repart.c:
```diff
@@ -5760,6 +5760,12 @@ static int context_minimize(Context *context) {
                 if (fstype_is_ro(p->format)) {
                         struct stat st;
 
+                        if (fd < 0) {
```
Is this check necessary? I think `assert(fd < 0);` is better.

yuwata commented on src/partition/repart.c:
```diff
@@ -5760,6 +5760,12 @@ static int context_minimize(Context *context) {
+                                        return log_error_errno(errno, "Failed to open temporary file %s: %m", temp);
+                        }
+
                         if (stat(temp, &st) < 0)
```
`fstat()` can be used?

poettering replied:
this should use fstat(), not stat()

yuwata commented on src/partition/repart.c:
```diff
@@ -5493,16 +5475,15 @@ static int resolve_copy_blocks_auto(
         if (found == 0)
-                return log_error_errno(SYNTHETIC_ERRNO(ENXIO),
-                                       "Unable to automatically discover suitable partition to copy blocks from.");
+                return 0;
```
When non-negative value returned, please initialize `ret_xyz`.

yuwata commented on man/repart.d.xml:
```diff
@@ -439,7 +434,9 @@
-        <para>This option cannot be combined with <varname>CopyBlocks=</varname>.</para>
+        <para>When this option is used in combination with <varname>CopyBlocks=</varname>,
+        <command>systemd-repart</command> will first try the <varname>CopyBlocks=</varname> logic. If that is
+        not possible, it falls back to the <varname>CopyFiles=</varname> logic.</para>
```
Maybe better to mention that `CopyFiles=` will be ignored when `CopyBlocks=` works.

yuwata commented on man/repart.d.xml:
```diff
@@ -566,6 +566,15 @@
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>Oem=</varname></term>
```
`OEM=` is better?? E.g. we have `DNS=` instead of `Dns=`.

yuwata commented on src/partition/repart.c:
```diff
@@ -6403,6 +6429,14 @@ static int parse_argv(int argc, char *argv[]) {
                         break;
 
+                case ARG_OEM:
+                        r = parse_boolean_argument("--oem=", optarg, NULL);
```
`NULL` -> `arg_oem`, then the assignment below can be dropped.

yuwata commented on test/units/testsuite-58.sh:
```diff
@@ -132,13 +132,15 @@ EOF
 Type=home
 Label=home-first
 Label=home-always-too-long-xxxxxxxxxxxxxx-%v
+Oem=yes
```
`Oem` -> `OEM`

yuwata replied:
Why the test succeeds??

DaanDeMeyer replied:
Improved the test so that it fails if OEM= is not enabled in this partition

poettering commented on src/partition/repart.c:
```diff
@@ -5387,19 +5382,6 @@ static int resolve_copy_blocks_auto(
-
-        if (restrict_devno == 0)
-                return log_error_errno(SYNTHETIC_ERRNO(EPERM),
-                                       "Automatic discovery of backing block devices not permitted in --root= mode, refusing.");
```
why drop this check?

poettering commented on src/partition/repart.c:
```diff
@@ -5492,17 +5474,13 @@ static int resolve_copy_blocks_auto(
 
-        if (found == 0)
-                return log_error_errno(SYNTHETIC_ERRNO(ENXIO),
-                                       "Unable to automatically discover suitable partition to copy blocks from.");
```
should be kept as a debug message

DaanDeMeyer commented:
@poettering This was reverted already please ignore

poettering commented on src/shared/creds-util.c:
```diff
@@ -253,6 +254,17 @@ int read_credential_strings_many_internal(
+
+        r = read_credential(name, &data, NULL);
+        if (r < 0)
+                return IN_SET(r, -ENXIO, -ENOENT) ? 0 : r;
```
i'd really leave this mangling to the caller. maybe some caller might want to default to true rather than false if unset
